server/mailet/mailetcontainer-camel/src/main/java/org/apache/james/mailetcontainer/impl/JamesMailSpooler.java

```diff
@@ -118,81 +100,50 @@ public void configure(HierarchicalConfiguration config) throws ConfigurationExce
+    private Mono<Void> handleOnQueueItem(MailQueueItem queueItem) {
+        TimeMetric timeMetric = metricFactory.timer(SPOOL_PROCESSING);
+        try {
+            processingActive.incrementAndGet();
+            return processMail(queueItem);
+        } catch (Throwable e) {
+            return Mono.error(e);
+        } finally {
+            processingActive.decrementAndGet();
+            timeMetric.stopAndPublish();
         }
     }
 
-    /**
-     * This routinely checks the message spool for messages, and processes them
-     * as necessary
-     */
-    @Override
-    public void run() {
-        LOGGER.info("Run {}: {}", getClass().getName(), Thread.currentThread().getName());
-        LOGGER.info("Queue={}", queue);
-
-        while (active.get()) {
-
-            final MailQueueItem queueItem;
-            try {
-                queueItem = queue.deQueue();
-                workerService.execute(() -> {
-                    TimeMetric timeMetric = metricFactory.timer(SPOOL_PROCESSING);
-                    try {
-                        numActive.incrementAndGet();
-
-                        // increase count
-                        processingActive.incrementAndGet();
-
-                        Mail mail = queueItem.getMail();
-                        LOGGER.debug("==== Begin processing mail {} ====", mail.getName());
-
-                        try {
-                            mailProcessor.service(mail);
-                            queueItem.done(true);
-                        } catch (Exception e) {
-                            if (active.get()) {
-                                LOGGER.error("Exception processing mail while spooling", e);
-                            }
-                            queueItem.done(false);
-
-                        } finally {
-                            LifecycleUtil.dispose(mail);
-                            mail = null;
-                        }
-                    } catch (Throwable e) {
-                        if (active.get()) {
-                            LOGGER.error("Exception processing mail while spooling", e);
-
-                        }
-                    } finally {
-                        processingActive.decrementAndGet();
-                        numActive.decrementAndGet();
-                        timeMetric.stopAndPublish();
-                    }
-
-                });
-            } catch (MailQueueException e1) {
-                if (active.get()) {
-                    LOGGER.error("Exception dequeue mail", e1);
-                }
-            } catch (InterruptedException interrupted) {
-                //MailSpooler is stopping
-            }
+    private Mono<Void> processMail(MailQueueItem queueItem) throws MailQueue.MailQueueException {
+        Mail mail = queueItem.getMail();
+        LOGGER.debug("==== Begin processing mail {} ====", mail.getName());
+        try {
+            mailProcessor.service(mail);
```

Comment:
Shouldn't it make sens to push that heavy operation in reactor and push the computation into it (Mono.fromRunnable?) rather than just using reactor as syntactic sugar once everything is done?

Reply:
Good question. In my understanding, processMail will be called on the flux and as the flux is published into the spooler threadpool, I assume each processMail is actually run in into a dedicated thread.
Returning a Mono that defers the computation would not add anything.
But I may interpret Reactor badly, feel free to correct me.

Reply:
Well that is also the way I understood it.

Just returning a Mono when everthing is done feels like sheating.